light/beacon/sync_committee.go

```diff
@@ -0,0 +1,547 @@
+			return c
+		} else {
+			log.Error("Missing serialized sync committee", "period", period, "committeeRoot", committeeRoot)
+		}
+	}
+	return nil
+}
+
+// EnforceForksAndConstraints rolls back committee updates that do not match the tracker's forks
+// and constraints and also starts new requests if possible (tracker mutex locked)
+func (s *SyncCommitteeTracker) EnforceForksAndConstraints() {
+	s.lock.Lock()
+	s.enforceForksAndConstraints()
+	s.lock.Unlock()
+}
+
+// enforceForksAndConstraints rolls back committee updates that do not match the tracker's forks
+// and constraints and also starts new requests if possible (tracker mutex expected)
+func (s *SyncCommitteeTracker) enforceForksAndConstraints() {
+	if !s.genesisInit || !s.chainInit {
+		return
+	}
+	s.committeeRootCache.Purge()
+	for s.nextPeriod > s.firstPeriod {
+		if update := s.GetBestUpdate(s.nextPeriod - 1); update == nil || s.verifyUpdate(update) { // check constraints and signature
+			if update == nil {
+				log.Error("Sync committee update missing", "period", s.nextPeriod-1)
+			}
+			break
+		}
+		s.nextPeriod--
+		s.deleteBestUpdate(s.nextPeriod)
+	}
+	if s.nextPeriod == s.firstPeriod {
+		if root, matchAll := s.constraints.CommitteeRoot(s.firstPeriod); matchAll || s.getSyncCommitteeRoot(s.firstPeriod) != root || s.getSyncCommittee(s.firstPeriod) == nil {
+			s.nextPeriod, s.firstPeriod, s.chainInit = 0, 0, false
+		}
+	}
+
+	s.retrySyncAllPeers()
+}
+
+// checkConstraints checks whether the signed headers of the given committee update is
+// on the right fork and the proven NextSyncCommitteeRoot matches the update chain constraints.
+func (s *SyncCommitteeTracker) checkConstraints(update *LightClientUpdate) bool {
+	if !s.genesisInit {
+		log.Error("SyncCommitteeTracker not initialized")
+		return false
+	}
+	root, matchAll := s.constraints.CommitteeRoot(uint64(update.Header.Slot)>>13 + 1)
+	return matchAll || root == update.NextSyncCommitteeRoot
+}
+
+// LightClientUpdate is a proof of the next sync committee root based on a header signed by the sync committee
+// of the given period. Optionally the update can prove quasi-finality by the signed header referring to a previous,
+// finalized header from the same period, and the finalized header referring to the next sync committee root.
+type LightClientUpdate struct {
+	Header                  Header
+	NextSyncCommitteeRoot   common.Hash
+	NextSyncCommitteeBranch MerkleValues
+	FinalizedHeader         Header
+	FinalityBranch          MerkleValues
+	SyncCommitteeBits       []byte
+	SyncCommitteeSignature  []byte
+	score                   UpdateScore
+}
+
+// Validate verifies the validity of the update
+func (update *LightClientUpdate) Validate() error {
+	if update.Header.Slot&0x1fff == 0x1fff {
+		// last slot of each period is not suitable for an update because it is signed by the next period's sync committee, proves the same committee it is signed by
+		return errors.New("Last slot of period")
```

Comment:
Why the last slot of period is always rejected? 

It can also be signed by the committee of next period which should be aware by light client already.

Reply:
The client is not aware of the next period committee when validating this proof since this is the proof that should prove it to the client. It proves nothing if it's signed by the same committee.